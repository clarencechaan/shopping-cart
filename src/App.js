import { BrowserRouter, Routes, Route } from "react-router-dom";
import NavBar from "./components/NavBar";
import Home from "./components/Home";
import MusicShop from "./components/MusicShop";
import MerchShop from "./components/MerchShop";
import SaleShop from "./components/SaleShop";
import Cart from "./components/Cart";
import Details from "./components/Details";
import items from "./items/items.js";
import "./styles/App.css";
import { useState, useEffect } from "react";

const App = () => {
  const [shopItems, setShopItems] = useState([]);
  const [cart, setCart] = useState([]);

  useEffect(() => setShopItems(items), []);

  function handleAddToCart(item, qty) {
    if (!qty) return;
    if (!isItemInCart(item, cart)) {
      setCart((prevCart) => {
        const newCartItem = { ...item, quantity: qty };
        return [...prevCart, newCartItem];
      });
    } else {
      setCart((prevCart) => {
        const prevCartItem = prevCart.find((c) => c.id === item.id);
        const newCartItem = {
          ...prevCartItem,
          quantity: prevCartItem.quantity + qty,
        };
        const index = prevCart.findIndex((c) => c.id === item.id);
        return [
          ...prevCart.slice(0, index),
          newCartItem,
          ...prevCart.slice(index + 1),
        ];
      });
    }
    removeZeroQuantity();
  }

  function removeZeroQuantity() {
    setCart((prevCart) => prevCart.filter((item) => item.quantity > 0));
  }

  function isItemInCart(item) {
    for (const c of cart) {
      if (c.id === item.id) return true;
    }
    return false;
  }

  function getQtyInCart(id) {
    if (isItemInCart({ id: id })) {
      return cart.find((c) => c.id === id).quantity;
    } else {
      return 0;
    }
  }

  function getTotalQtyInCart() {
    let total = 0;
    for (const c of cart) {
      total += c.quantity;
    }
    return total;
  }

  // function addQty(id, qty) {
  //   setCart((prevCart) => {
  //     const prevCartItem = prevCart.find((c) => c.id === id);
  //     const newCartItem = {
  //       ...prevCartItem,
  //       quantity: prevCartItem.quantity + 1,
  //     };
  //     const index = prevCart.findIndex((c) => c.id === id);
  //     return [
  //       ...prevCart.slice(0, index),
  //       newCartItem,
  //       ...prevCart.slice(index + 1),
  //     ];
  //   });
  // }

  console.log(cart);

  return (
    <BrowserRouter>
      <NavBar totalQtyInCart={getTotalQtyInCart()} />
      <Routes>
        <Route path="/" element={<Home />} />
        <Route
          path="/music-shop"
          element={<MusicShop shopItems={shopItems} />}
        />
        <Route
          path="/merch-shop"
          element={<MerchShop shopItems={shopItems} />}
        />
        <Route path="/sale-shop" element={<SaleShop shopItems={shopItems} />} />
        <Route
          path="/cart"
          element={
            <Cart
              cartItems={cart}
              getTotalQtyInCart={getTotalQtyInCart}
              handleAddToCart={handleAddToCart}
            />
          }
        />
        <Route
          path="/details"
          element={
            <Details
              handleAddToCart={handleAddToCart}
              getQtyInCart={getQtyInCart}
            />
          }
        />
      </Routes>
    </BrowserRouter>
  );
};

export default App;
