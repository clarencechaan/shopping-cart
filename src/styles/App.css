body {
  margin: 0;
}

#root {
  width: 100vw;
  height: 100vh;
  max-width: 100%;
  font-family: "Pavanam", sans-serif;
  display: flex;
  flex-direction: column;
  align-items: center;
}

#root > div:last-child {
  flex: 1;
}

.home {
  width: 100%;
  height: 100%;
}

.music-shop,
.merch-shop,
.sale-shop,
.cart {
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  width: 1192px;
  padding: 0 24px 24px 24px;
}

.cart {
  width: 1096px;
}
