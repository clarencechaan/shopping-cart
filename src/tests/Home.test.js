import React from "react";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import App from "../App";

describe("Home component", () => {
  it("renders Home", () => {
    const { container } = render(<App />);

    expect(container).toMatchSnapshot();
  });

  it("Home link music", () => {
    const { container } = render(<App />);

    const button = screen.getByRole("button", {
      name: /^SHOP MUSIC$/,
    });
    userEvent.click(button);

    expect(container).toMatchSnapshot();
  });

  it("Home link merch", () => {
    const { container } = render(<App />);
    const home = screen.getByText(/HOME/i);
    userEvent.click(home);
    const button = screen.getByRole("button", {
      name: /^SHOP MERCH$/,
    });
    userEvent.click(button);

    expect(container).toMatchSnapshot();
  });
});
