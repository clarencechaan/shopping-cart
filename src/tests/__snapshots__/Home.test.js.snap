// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home component Home link merch 1`] = `
<div>
  <nav
    class="nav-bar"
  >
    <a
      class="logo"
      href="/shopping-cart"
    >
      auditty
    </a>
    <div>
      <a
        class="nav-link"
        href="/shopping-cart"
      >
        HOME
      </a>
      <a
        class="nav-link"
        href="/shopping-cart/music-shop"
      >
        MUSIC
      </a>
      <a
        class="nav-link"
        href="/shopping-cart/merch-shop"
      >
        MERCH
      </a>
      <a
        class="nav-link"
        href="/shopping-cart/sale-shop"
        style="color: red;"
      >
        SALE
      </a>
    </div>
    <a
      class="cart-logo"
      href="/shopping-cart/cart"
    >
      <img
        alt="cart"
        src="shopping-cart.png"
      />
      <div
        class="cart-badge hidden"
      />
    </a>
  </nav>
  <div
    class="merch-shop"
  >
    <h2>
      Merch
    </h2>
    <div
      class="items-container"
    >
      <div
        class="shop-item"
      >
        <a
          href="/shopping-cart/details"
        >
          <img
            alt="OVO Essentials Hoodie"
            class="shop-item-img"
            src="ovo-hoodie.webp"
          />
        </a>
        <div
          class="shop-title"
        >
          OVO Essentials Hoodie
        </div>
        <div
          class="shop-price"
        >
          $
          79.99
        </div>
      </div>
      <div
        class="shop-item"
      >
        <a
          href="/shopping-cart/details"
        >
          <img
            alt="Macadelic 2012 Tour Tee"
            class="shop-item-img"
            src="macadelic-tee.webp"
          />
        </a>
        <div
          class="shop-title"
        >
          Macadelic 2012 Tour Tee
        </div>
        <div
          class="shop-price"
        >
          $
          29.99
        </div>
      </div>
      <div
        class="shop-item"
      >
        <a
          href="/shopping-cart/details"
        >
          <img
            alt="Kanye West “Holy Spirit” Sweatshirt"
            class="shop-item-img"
            src="kanye-holy-spirit.jpeg"
          />
        </a>
        <div
          class="shop-title"
        >
          Kanye West “Holy Spirit” Sweatshirt
        </div>
        <div
          class="shop-price"
        >
          $
          59.99
        </div>
      </div>
      <div
        class="shop-item"
      >
        <a
          href="/shopping-cart/details"
        >
          <img
            alt="\\"Expensive Pain\\" Hoodie"
            class="shop-item-img"
            src="expensive-pain-hoodie.webp"
          />
        </a>
        <div
          class="shop-title"
        >
          "Expensive Pain" Hoodie
        </div>
        <div
          class="shop-price"
        >
          $
          79.99
        </div>
      </div>
      <div
        class="shop-item"
      >
        <a
          href="/shopping-cart/details"
        >
          <img
            alt="Pivot Gang x TOMBOGO Bucket Hat"
            class="shop-item-img"
            src="pivot-gang-hat.webp"
          />
        </a>
        <div
          class="shop-title"
        >
          Pivot Gang x TOMBOGO Bucket Hat
        </div>
        <div
          class="shop-price"
        >
          $
          29.99
        </div>
      </div>
      <div
        class="shop-item"
      >
        <a
          href="/shopping-cart/details"
        >
          <img
            alt="Drogas Wave Album Hoodie"
            class="shop-item-img"
            src="drogas-hoodie.webp"
          />
        </a>
        <div
          class="shop-title"
        >
          Drogas Wave Album Hoodie
        </div>
        <div
          class="shop-price"
        >
          $
          69.99
        </div>
      </div>
      <div
        class="shop-item"
      >
        <a
          href="/shopping-cart/details"
        >
          <img
            alt="ISSA Money Pool Tee"
            class="shop-item-img"
            src="issa-tee.jpeg"
          />
        </a>
        <div
          class="shop-title"
        >
          ISSA Money Pool Tee
        </div>
        <div
          class="shop-price"
        >
          $
          24.99
        </div>
      </div>
      <div
        class="shop-item"
      >
        <a
          href="/shopping-cart/details"
        >
          <img
            alt="Pro Era 47 Shorts"
            class="shop-item-img"
            src="pro-era-shorts.webp"
          />
        </a>
        <div
          class="shop-title"
        >
          Pro Era 47 Shorts
        </div>
        <div
          class="shop-price"
        >
          $
          44.99
        </div>
      </div>
      <div
        class="shop-item"
      >
        <a
          href="/shopping-cart/details"
        >
          <img
            alt="VLONE x Saint Mxxxxxx Sweat Pants "
            class="shop-item-img"
            src="vlone-pants.webp"
          />
        </a>
        <div
          class="shop-title"
        >
          VLONE x Saint Mxxxxxx Sweat Pants 
        </div>
        <div
          class="shop-price"
        >
          $
          64.99
        </div>
      </div>
      <div
        class="shop-item"
      >
        <a
          href="/shopping-cart/details"
        >
          <img
            alt="Chance 3 New Era Cap"
            class="shop-item-img"
            src="chance-hat.webp"
          />
        </a>
        <div
          class="shop-title"
        >
          Chance 3 New Era Cap
        </div>
        <div
          class="shop-price"
        >
          $
          34.99
        </div>
      </div>
      <div
        class="shop-item"
      >
        <a
          href="/shopping-cart/details"
        >
          <img
            alt="Enjoy Life Tee"
            class="shop-item-img"
            src="6lack-tee.webp"
          />
        </a>
        <div
          class="shop-title"
        >
          Enjoy Life Tee
        </div>
        <div
          class="shop-price"
        >
          $
          24.99
        </div>
      </div>
      <div
        class="shop-item"
      >
        <a
          href="/shopping-cart/details"
        >
          <img
            alt="TDE Hoodie"
            class="shop-item-img"
            src="tde-hoodie.webp"
          />
        </a>
        <div
          class="shop-title"
        >
          TDE Hoodie
        </div>
        <div
          class="shop-price"
        >
          $
          64.99
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Home component Home link music 1`] = `
<div>
  <nav
    class="nav-bar"
  >
    <a
      class="logo"
      href="/shopping-cart"
    >
      auditty
    </a>
    <div>
      <a
        class="nav-link"
        href="/shopping-cart"
      >
        HOME
      </a>
      <a
        class="nav-link"
        href="/shopping-cart/music-shop"
      >
        MUSIC
      </a>
      <a
        class="nav-link"
        href="/shopping-cart/merch-shop"
      >
        MERCH
      </a>
      <a
        class="nav-link"
        href="/shopping-cart/sale-shop"
        style="color: red;"
      >
        SALE
      </a>
    </div>
    <a
      class="cart-logo"
      href="/shopping-cart/cart"
    >
      <img
        alt="cart"
        src="shopping-cart.png"
      />
      <div
        class="cart-badge hidden"
      />
    </a>
  </nav>
  <div
    class="music-shop"
  >
    <h2>
      Music
    </h2>
    <div
      class="items-container"
    >
      <div
        class="shop-item"
      >
        <a
          href="/shopping-cart/details"
        >
          <img
            alt="My Beautiful Dark Twisted Fantasy by Kanye West"
            class="shop-item-img"
            src="mbdtf.jpg"
          />
        </a>
        <div
          class="shop-title"
        >
          My Beautiful Dark Twisted Fantasy by Kanye West
        </div>
        <div
          class="shop-price"
        >
          $
          11.99
        </div>
      </div>
      <div
        class="shop-item"
      >
        <a
          href="/shopping-cart/details"
        >
          <img
            alt="Good Kid, m.A.A.d City by Kendrick Lamar"
            class="shop-item-img"
            src="gkmc.jpg"
          />
        </a>
        <div
          class="shop-title"
        >
          Good Kid, m.A.A.d City by Kendrick Lamar
        </div>
        <div
          class="shop-price"
        >
          $
          11.99
        </div>
      </div>
      <div
        class="shop-item"
      >
        <a
          href="/shopping-cart/details"
        >
          <img
            alt="Take Care by Drake"
            class="shop-item-img"
            src="take-care.jpg"
          />
        </a>
        <div
          class="shop-title"
        >
          Take Care by Drake
        </div>
        <div
          class="shop-price"
        >
          $
          9.99
        </div>
      </div>
      <div
        class="shop-item"
      >
        <a
          href="/shopping-cart/details"
        >
          <img
            alt="TA13OO by Denzel Curry"
            class="shop-item-img"
            src="taboo.jpg"
          />
        </a>
        <div
          class="shop-title"
        >
          TA13OO by Denzel Curry
        </div>
        <div
          class="shop-price"
        >
          $
          11.99
        </div>
      </div>
      <div
        class="shop-item"
      >
        <a
          href="/shopping-cart/details"
        >
          <img
            alt="B4.DA.$$ by Joey Bada$$"
            class="shop-item-img"
            src="badass.jpg"
          />
        </a>
        <div
          class="shop-title"
        >
          B4.DA.$$ by Joey Bada$$
        </div>
        <div
          class="shop-price"
        >
          $
          11.99
        </div>
      </div>
      <div
        class="shop-item"
      >
        <a
          href="/shopping-cart/details"
        >
          <img
            alt="Channel Orange by Frank Ocean"
            class="shop-item-img"
            src="channel-orange.jpg"
          />
        </a>
        <div
          class="shop-title"
        >
          Channel Orange by Frank Ocean
        </div>
        <div
          class="shop-price"
        >
          $
          11.99
        </div>
      </div>
      <div
        class="shop-item"
      >
        <a
          href="/shopping-cart/details"
        >
          <img
            alt="Care for Me by Saba"
            class="shop-item-img"
            src="care.jpg"
          />
        </a>
        <div
          class="shop-title"
        >
          Care for Me by Saba
        </div>
        <div
          class="shop-price"
        >
          $
          9.99
        </div>
      </div>
      <div
        class="shop-item"
      >
        <a
          href="/shopping-cart/details"
        >
          <img
            alt="Food & Liquor by Lupe Fiasco"
            class="shop-item-img"
            src="food-liquor.jpg"
          />
        </a>
        <div
          class="shop-title"
        >
          Food & Liquor by Lupe Fiasco
        </div>
        <div
          class="shop-price"
        >
          $
          9.99
        </div>
      </div>
      <div
        class="shop-item"
      >
        <a
          href="/shopping-cart/details"
        >
          <img
            alt="Man on the Moon: The End of Day by Kid Cudi"
            class="shop-item-img"
            src="motm.jpg"
          />
        </a>
        <div
          class="shop-title"
        >
          Man on the Moon: The End of Day by Kid Cudi
        </div>
        <div
          class="shop-price"
        >
          $
          11.99
        </div>
      </div>
      <div
        class="shop-item"
      >
        <a
          href="/shopping-cart/details"
        >
          <img
            alt="Dreams Worth More Than Money by Meek Mill"
            class="shop-item-img"
            src="dwmtm.jpg"
          />
        </a>
        <div
          class="shop-title"
        >
          Dreams Worth More Than Money by Meek Mill
        </div>
        <div
          class="shop-price"
        >
          $
          11.99
        </div>
      </div>
      <div
        class="shop-item"
      >
        <a
          href="/shopping-cart/details"
        >
          <img
            alt="Savage Mode by 21 Savage"
            class="shop-item-img"
            src="savage.jpg"
          />
        </a>
        <div
          class="shop-title"
        >
          Savage Mode by 21 Savage
        </div>
        <div
          class="shop-price"
        >
          $
          11.99
        </div>
      </div>
      <div
        class="shop-item"
      >
        <a
          href="/shopping-cart/details"
        >
          <img
            alt="Swimming by Mac Miller"
            class="shop-item-img"
            src="swimming.jpg"
          />
        </a>
        <div
          class="shop-title"
        >
          Swimming by Mac Miller
        </div>
        <div
          class="shop-price"
        >
          $
          11.99
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Home component renders Home 1`] = `
<div>
  <nav
    class="nav-bar"
  >
    <a
      class="logo"
      href="/shopping-cart"
    >
      auditty
    </a>
    <div>
      <a
        class="nav-link"
        href="/shopping-cart"
      >
        HOME
      </a>
      <a
        class="nav-link"
        href="/shopping-cart/music-shop"
      >
        MUSIC
      </a>
      <a
        class="nav-link"
        href="/shopping-cart/merch-shop"
      >
        MERCH
      </a>
      <a
        class="nav-link"
        href="/shopping-cart/sale-shop"
        style="color: red;"
      >
        SALE
      </a>
    </div>
    <a
      class="cart-logo"
      href="/shopping-cart/cart"
    >
      <img
        alt="cart"
        src="shopping-cart.png"
      />
      <div
        class="cart-badge hidden"
      />
    </a>
  </nav>
  <div
    class="home"
  >
    <div
      class="shop-music"
    >
      <a
        class="shop-button-link"
        href="/shopping-cart/music-shop"
      >
        <button
          class="shop-button"
        >
          SHOP MUSIC
        </button>
      </a>
      <div
        class="shop-music-background"
        style="background-image: url(record.jpg); background-size: cover; background-position: center;"
      >
        <div
          class="white-overlay"
        />
      </div>
    </div>
    <div
      class="shop-merch"
    >
      <a
        class="shop-button-link"
        href="/shopping-cart/merch-shop"
      >
        <button
          class="shop-button"
        >
          SHOP MERCH
        </button>
      </a>
      <div
        class="shop-merch-background"
        style="background-image: url(merch.jpg); background-size: cover; background-position: center;"
      >
        <div
          class="white-overlay"
        />
      </div>
    </div>
     
  </div>
</div>
`;
